File: src/api/schemas/csv_sql.py
"""API schemas for CSV-SQL operations."""
from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


# ==================== Column Schema ====================


class ColumnInfo(BaseModel):
    """Information about a table column."""

    name: str
    type: str
    original: Optional[str] = None


# ==================== Table Schemas ====================


class CSVTableResponse(BaseModel):
    """Response containing CSV table details."""

    table_name: str
    original_filename: Optional[str] = None
    row_count: Optional[int] = None
    column_count: Optional[int] = None
    columns: Optional[List[ColumnInfo]] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class CSVTableListResponse(BaseModel):
    """Response containing list of CSV tables."""

    tables: List[CSVTableResponse]
    total: int


# ==================== Upload Schemas ====================


class UploadCSVRequest(BaseModel):
    """Request to upload a CSV file."""

    table_name: str = Field(..., min_length=1, max_length=100, description="Name for the table")
    if_exists: str = Field(
        "fail",
        pattern="^(fail|replace|append)$",
        description="What to do if table exists: fail, replace, or append",
    )


class UploadCSVResponse(BaseModel):
    """Response after uploading a CSV file."""

    success: bool
    table_name: Optional[str] = None
    schema: Optional[str] = None
    full_path: Optional[str] = None
    row_count: Optional[int] = None
    column_count: Optional[int] = None
    columns: Optional[List[ColumnInfo]] = None
    message: Optional[str] = None
    error: Optional[str] = None


# ==================== Query Schemas ====================


class ExecuteSQLRequest(BaseModel):
    """Request to execute a SQL query."""

    query: str = Field(..., min_length=1, description="SQL query to execute")
    limit: int = Field(100, ge=1, le=10000, description="Maximum rows to return")


class ExecuteSQLResponse(BaseModel):
    """Response after executing a SQL query."""

    success: bool
    query_type: Optional[str] = None
    row_count: Optional[int] = None
    affected_rows: Optional[int] = None
    results: Optional[str] = None
    query: Optional[str] = None
    message: Optional[str] = None
    error: Optional[str] = None


# ==================== Schema Response ====================


class TableSchemaResponse(BaseModel):
    """Response containing table schema."""

    success: bool
    table_name: Optional[str] = None
    schema: Optional[str] = None
    full_path: Optional[str] = None
    columns: Optional[List[ColumnInfo]] = None
    row_count: Optional[int] = None
    original_filename: Optional[str] = None
    created_at: Optional[str] = None
    error: Optional[str] = None


# ==================== Sample Data Response ====================


class SampleDataResponse(BaseModel):
    """Response containing sample data from a table."""

    success: bool
    table_name: Optional[str] = None
    sample: Optional[str] = None
    row_count: Optional[int] = None
    error: Optional[str] = None


# ==================== Delete Response ====================


class DeleteTableResponse(BaseModel):
    """Response after deleting a table."""

    success: bool
    message: Optional[str] = None
    error: Optional[str] = None

File: web/lib/api/csv-sql.ts
/**
 * CSV-SQL API client.
 *
 * Provides functions for uploading CSV files and querying them via SQL.
 */

import { apiClient } from "./client";

// ==================== Types ====================

export interface ColumnInfo {
  name: string;
  type: string;
  original?: string;
}

export interface CSVTable {
  table_name: string;
  original_filename?: string;
  row_count?: number;
  column_count?: number;
  columns?: ColumnInfo[];
  created_at?: string;
  updated_at?: string;
}

export interface CSVTableListResponse {
  tables: CSVTable[];
  total: number;
}

export interface UploadCSVResponse {
  success: boolean;
  table_name?: string;
  schema?: string;
  full_path?: string;
  row_count?: number;
  column_count?: number;
  columns?: ColumnInfo[];
  message?: string;
  error?: string;
}

export interface TableSchemaResponse {
  success: boolean;
  table_name?: string;
  schema?: string;
  full_path?: string;
  columns?: ColumnInfo[];
  row_count?: number;
  original_filename?: string;
  created_at?: string;
  error?: string;
}

export interface SampleDataResponse {
  success: boolean;
  table_name?: string;
  sample?: string;
  row_count?: number;
  error?: string;
}

export interface ExecuteSQLResponse {
  success: boolean;
  query_type?: string;
  row_count?: number;
  affected_rows?: number;
  results?: string;
  query?: string;
  message?: string;
  error?: string;
}

export interface DeleteTableResponse {
  success: boolean;
  message?: string;
  error?: string;
}

// ==================== API Functions ====================

/**
 * Upload a CSV file and create a PostgreSQL table.
 */
export async function uploadCSV(
  file: File,
  tableName: string,
  ifExists: "fail" | "replace" | "append" = "fail"
): Promise<UploadCSVResponse> {
  const formData = new FormData();
  formData.append("file", file);
  formData.append("table_name", tableName);
  formData.append("if_exists", ifExists);

  const response = await apiClient.post<UploadCSVResponse>(
    "/csv-sql/upload",
    formData,
    {
      headers: {
        "Content-Type": "multipart/form-data",
      },
    }
  );
  return response.data;
}

/**
 * List all CSV tables for the current user.
 */
export async function listCSVTables(): Promise<CSVTableListResponse> {
  const response = await apiClient.get<CSVTableListResponse>("/csv-sql/tables");
  return response.data;
}

/**
 * Get the schema for a specific table.
 */
export async function getTableSchema(
  tableName: string
): Promise<TableSchemaResponse> {
  const response = await apiClient.get<TableSchemaResponse>(
    `/csv-sql/tables/${encodeURIComponent(tableName)}/schema`
  );
  return response.data;
}

/**
 * Get sample data from a table.
 */
export async function getSampleData(
  tableName: string,
  rows: number = 5
): Promise<SampleDataResponse> {
  const response = await apiClient.get<SampleDataResponse>(
    `/csv-sql/tables/${encodeURIComponent(tableName)}/sample`,
    { params: { rows } }
  );
  return response.data;
}

/**
 * Execute a SQL query on the CSV tables.
 */
export async function executeSQL(
  query: string,
  limit: number = 100
): Promise<ExecuteSQLResponse> {
  const response = await apiClient.post<ExecuteSQLResponse>("/csv-sql/query", {
    query,
    limit,
  });
  return response.data;
}

/**
 * Delete a CSV table.
 */
export async function deleteCSVTable(
  tableName: string
): Promise<DeleteTableResponse> {
  const response = await apiClient.delete<DeleteTableResponse>(
    `/csv-sql/tables/${encodeURIComponent(tableName)}`
  );
  return response.data;
}
